sys/stdvga.c

```diff
@@ -0,0 +1,165 @@
+
+/* Detailed information about VGA registers is available at
+   http://www.osdever.net/FreeVGA/vga/vga.htm */
+
+#define VGA_PALETTE_ADDR 0x3C8
+#define VGA_PALETTE_DATA 0x3C9
+#define VGA_AR_ADDR 0x3C0
+#define VGA_AR_PAS 0x20 /* Palette Address Source bit */
+
+/* Bochs VBE. Simplifies VGA graphics mode configuration a great deal. Some
+   documentation is available at http://wiki.osdev.org/Bochs_VBE_Extensions */
+#define VBE_DISPI_INDEX_XRES 0x01
+#define VBE_DISPI_INDEX_YRES 0x02
+#define VBE_DISPI_INDEX_BPP 0x03
+#define VBE_DISPI_INDEX_ENABLE 0x04
+#define VBE_DISPI_ENABLED 0x01 /* VBE Enabled bit */
+
+/* Offsets for accessing ioports via PCI BAR1 (MMIO) */
+#define VGA_MMIO_OFFSET (0x400 - 0x3c0)
+#define VBE_MMIO_OFFSET 0x500
+
+/* The general overview of the QEMU std vga device is available at
+   https://github.com/qemu/qemu/blob/master/docs/specs/standard-vga.txt */
+#define VGA_QEMU_STDVGA_VENDOR_ID 0x1234
+#define VGA_QEMU_STDVGA_DEVICE_ID 0x1111
+
+static void stdvga_io_write(stdvga_device_t *vga, uint16_t reg, uint8_t value) {
+  bus_space_write_1(&vga->io, reg + VGA_MMIO_OFFSET, value);
+}
+static uint8_t stdvga_io_read(stdvga_device_t *vga, uint16_t reg)
+  __attribute__((unused));
+static uint8_t stdvga_io_read(stdvga_device_t *vga, uint16_t reg) {
+  return bus_space_read_1(&vga->io, reg + VGA_MMIO_OFFSET);
+}
+static void stdvga_vbe_write(stdvga_device_t *vga, uint16_t reg,
+                             uint16_t value) {
+  /* <<1 shift enables access to 16-bit registers. */
+  bus_space_write_2(&vga->io, (reg << 1) + VBE_MMIO_OFFSET, value);
+}
+static uint16_t stdvga_vbe_read(stdvga_device_t *vga, uint16_t reg) {
+  return bus_space_read_2(&vga->io, (reg << 1) + VBE_MMIO_OFFSET);
+}
+
+static void stdvga_palette_write_single(stdvga_device_t *stdvga, uint8_t offset,
+                                        uint8_t r, uint8_t g, uint8_t b) {
+  stdvga_io_write(stdvga, VGA_PALETTE_ADDR, offset);
+  stdvga_io_write(stdvga, VGA_PALETTE_DATA, r >> 2);
+  stdvga_io_write(stdvga, VGA_PALETTE_DATA, g >> 2);
+  stdvga_io_write(stdvga, VGA_PALETTE_DATA, b >> 2);
+}
+
+static void stdvga_palette_write_buffer(stdvga_device_t *stdvga,
+                                        const uint8_t buf[VGA_PALETTE_SIZE]) {
+  for (int i = 0; i < VGA_PALETTE_SIZE / 3; i++)
+    stdvga_palette_write_single(stdvga, i, buf[3 * i + 0], buf[3 * i + 1],
+                                buf[3 * i + 2]);
+}
+
+static int stdvga_palette_write(vga_device_t *vga, uio_t *uio) {
+  stdvga_device_t *stdvga = STDVGA_FROM_VGA(vga);
+  int error = uiomove_frombuf(stdvga->palette_buffer, VGA_PALETTE_SIZE, uio);
+  if (error)
+    return error;
+  /* TODO: Only update the modified area. */
+  stdvga_palette_write_buffer(stdvga, stdvga->palette_buffer);
+  return 0;
+}
+
+static int stdvga_fb_write(vga_device_t *vga, uio_t *uio) {
+  stdvga_device_t *stdvga = STDVGA_FROM_VGA(vga);
+  /* TODO: I'd love to pass the uio to the bus directly, instead of using a
```

Comment:
[bus_space_map](https://www.freebsd.org/cgi/man.cgi?query=bus_space_map&sektion=9) call from *BSDs allows to map `RT_MEMORY` resource into kernel (and user?) virtual address space. Some day we will implement it and use BUS_SPACE_MAP_LINEAR for frambuffer memory.

> **BUS_SPACE_MAP_LINEAR** Try to map the space so that its contents can be	accessed linearly via normal memory access methods (e.g. pointer dereferencing and structure accesses).  This is useful when software wants to do direct access to a memory device, e.g. a frame buffer. [...]